File: 295_Find_Median_from_Data_Stream.cpp
class MedianFinder {
public:
    /** initialize your data structure here. */
    priority_queue<long> small, big;
    
    MedianFinder() {
    }
    
    void addNum(int num) {
        small.push(num);
        big.push(-small.top());
        small.pop();
        if (small.size() < big.size()) {
            small.push(-big.top());
            big.pop();
        }
    }
    
    double findMedian() {
        return small.size() > big.size() ? small.top() : (small.top() - big.top()) / 2.0;
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */

File: 295_Find_Median_from_Data_Stream.py
class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.bigger = []
        self.smaller = []

    def addNum(self, num: int) -> None:
        heapq.heappush(self.bigger, num)
        heapq.heappush(self.smaller, -heapq.heappop(self.bigger))
        if len(self.bigger) < len(self.smaller):
            heapq.heappush(self.bigger, -heapq.heappop(self.smaller))
    def findMedian(self) -> float:
        if len(self.bigger) > len(self.smaller):
            return self.bigger[0]
        else:
            return (self.bigger[0] - self.smaller[0]) / 2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()
